thrust/thrust/system/cuda/detail/temporary_buffer.h

```diff
@@ -26,4 +26,107 @@
 #  pragma system_header
 #endif // no system header
 
-// this system has no special temporary buffer functions
+#include <thrust/detail/type_traits/pointer_traits.h>
+#include <thrust/system/cuda/detail/par.h>
+#include <thrust/system/detail/bad_alloc.h>
+
+THRUST_NAMESPACE_BEGIN
+namespace cuda_cub
+{
+
+// If par_nosync does not have a user provided allocator attached, these
+// overloads should be selected.
+
+template <typename T>
+_CCCL_HOST pair<T*, ::cuda::std::ptrdiff_t> get_temporary_buffer(par_nosync_t&, ::cuda::std::ptrdiff_t n)
+{
+  void* ptr;
+  cudaError_t status = cudaMallocAsync(&ptr, sizeof(T) * n, nullptr);
+
+  if (status != cudaSuccess)
+  {
+    cudaGetLastError(); // Clear the CUDA global error state.
+
+    // That didn't work. We could be somewhere where async allocation isn't
+    // supported like Windows, so try again with cudaMalloc.
+    status = cudaMalloc(&ptr, sizeof(T) * n);
+
+    if (status != cudaSuccess)
+    {
+      throw system::detail::bad_alloc(cuda_category().message(status).c_str());
+    }
+  }
+
+  return make_pair(reinterpret_pointer_cast<T*>(ptr), n);
+}
+
+template <typename Pointer>
+_CCCL_HOST void return_temporary_buffer(par_nosync_t&, Pointer ptr, ::cuda::std::ptrdiff_t n)
+{
+  void* void_ptr = reinterpret_pointer_cast<void*>(ptr);
```

Comment:
In libcu++ we have `__voidify` which does the proper cast to void thingy

Reply:
I think the point here is to cast a `thrust::device_pointer<T>` into it's underlying `T*`. Since `T*` converts to `void*` automatically, we should be able to just `thrust::raw_pointer_cast(ptr)` here.

Reply:
Bernhard is right, the intent is to go from `device_pointer` to `T*`. `reinterpert_pointer_cast` is what we use elsewhere in `*_temporary_buffer` machinery which is why we used it. I also think he's right that we can just use `raw_pointer_cast` here.